# SF-DAT-21 | Unit Project 3

In this project, you will perform a logistic regression on the admissions data we've been working with in Unit Projects 1 and 2.

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import statsmodels.api as sm
import pylab as pl
%matplotlib inline

In [3]:
df_raw = pd.read_csv("../../dataset/admissions.csv")
df = df_raw.dropna()
print(df.head())

   admit    gre   gpa  prestige
0      0  380.0  3.61       3.0
1      1  660.0  3.67       3.0
2      1  800.0  4.00       1.0
3      1  640.0  3.19       4.0
4      0  520.0  2.93       4.0


## Part 1. Frequency Tables

#### Question 1. Let's create a frequency table of our variables.

In [4]:
# frequency table for prestige and whether or not someone was admitted

prestige = df['prestige']
print('Prestige Frequency Table:')
prestige.value_counts()

Prestige Frequency Table:


2.0    148
3.0    121
4.0     67
1.0     61
Name: prestige, dtype: int64

In [5]:
admit = df['admit']
print('Admit Frequency Table:')
admit.value_counts()

Admit Frequency Table:


0    271
1    126
Name: admit, dtype: int64

## Part 2. Return of dummy variables

#### Question 2.1. Create class or dummy variables for prestige.

In [6]:
prestige_dummy_df = pd.get_dummies(df['prestige'], prefix = 'Prestige')
prestige_dummy_df

Prestige_1.0  Prestige_2.0  Prestige_3.0  Prestige_4.0
0             0.0           0.0           1.0           0.0
1             0.0           0.0           1.0           0.0
2             1.0           0.0           0.0           0.0
3             0.0           0.0           0.0           1.0
4             0.0           0.0           0.0           1.0
5             0.0           1.0           0.0           0.0
6             1.0           0.0           0.0           0.0
7             0.0           1.0           0.0           0.0
8             0.0           0.0           1.0           0.0
9             0.0           1.0           0.0           0.0
10            0.0           0.0           0.0           1.0
11            1.0           0.0           0.0           0.0
12            1.0           0.0           0.0           0.0
13            0.0           1.0           0.0           0.0
14            1.0           0.0           0.0           0.0
15            0.0           0.0           1.0           0.0
16            0.0           0.0           0.0           1.0
17            0.0           0.0           1.0           0.0
18            0.0           1.0           0.0           0.0
19            1.0           0.0           0.0           0.0
20            0.0           0.0           1.0           0.0
21            0.0           1.0           0.0           0.0
22            0.0           0.0           0.0           1.0
23            0.0           0.0           0.0           1.0
24            0.0           1.0           0.0           0.0
25            1.0           0.0           0.0           0.0
26            1.0           0.0           0.0           0.0
27            0.0           0.0           0.0           1.0
28            0.0           1.0           0.0           0.0
29            1.0           0.0           0.0           0.0
..            ...           ...           ...           ...
370           0.0           1.0           0.0           0.0
371           0.0           0.0           1.0           0.0
372           1.0           0.0           0.0           0.0
373           1.0           0.0           0.0           0.0
374           0.0           1.0           0.0           0.0
375           0.0           0.0           0.0           1.0
376           0.0           1.0           0.0           0.0
377           0.0           1.0           0.0           0.0
378           0.0           0.0           1.0           0.0
379           0.0           1.0           0.0           0.0
380           0.0           1.0           0.0           0.0
381           0.0           1.0           0.0           0.0
382           0.0           1.0           0.0           0.0
383           1.0           0.0           0.0           0.0
384           0.0           1.0           0.0           0.0
385           1.0           0.0           0.0           0.0
386           0.0           1.0           0.0           0.0
387           0.0           1.0           0.0           0.0
388           0.0           1.0           0.0           0.0
389           0.0           1.0           0.0           0.0
390           0.0           1.0           0.0           0.0
391           0.0           1.0           0.0           0.0
392           0.0           0.0           1.0           0.0
393           0.0           1.0           0.0           0.0
394           0.0           0.0           1.0           0.0
395           0.0           1.0           0.0           0.0
396           0.0           0.0           1.0           0.0
397           0.0           1.0           0.0           0.0
398           0.0           1.0           0.0           0.0
399           0.0           0.0           1.0           0.0

[397 rows x 4 columns]

#### Question 2.2. When modeling our class variables, how many do we need?

Answer: We need four dummies for prestige, since schools can have a prestige rating of 1, 2, 3, or 4. We need two variables for admit because it is a binary outcome (each applicant is either admitted or rejected).

## Part 3. Hand calculating odds ratios

Develop your intuition about expected outcomes by hand calculating odds ratios.

In [7]:
cols_to_keep = ['admit', 'gre', 'gpa']
handCalc = df[cols_to_keep].join(prestige_dummy_df.ix[:, 'Prestige_1.0':])
print(handCalc.head())

   admit    gre   gpa  Prestige_1.0  Prestige_2.0  Prestige_3.0  Prestige_4.0
0      0  380.0  3.61           0.0           0.0           1.0           0.0
1      1  660.0  3.67           0.0           0.0           1.0           0.0
2      1  800.0  4.00           1.0           0.0           0.0           0.0
3      1  640.0  3.19           0.0           0.0           0.0           1.0
4      0  520.0  2.93           0.0           0.0           0.0           1.0


In [25]:
# crosstab of Prestige 1.0 and admission decision
pd.crosstab(index = ['admit'], columns = handCalc['Prestige_1.0'])

Prestige_1.0  0.0  1.0
row_0                 
admit         336   61

In [100]:
# frequency table cutting prestige and whether or not someone was admitted
ft = pd.crosstab(index = [handCalc['Prestige_1.0'], handCalc['Prestige_2.0'], handCalc['Prestige_3.0'], handCalc['Prestige_4.0']],
            columns = handCalc['admit'], rownames = ['P1', 'P2', 'P3', 'P4'], margins = True)
ft

admit              0    1  All
P1  P2  P3  P4                
0.0 0.0 0.0 1.0   55   12   67
        1.0 0.0   93   28  121
    1.0 0.0 0.0   95   53  148
1.0 0.0 0.0 0.0   28   33   61
All              271  126  397

#### Question 3.1. Use the cross tab above to calculate the odds of being admitted to grad school if you attended a #1 ranked college.

In [74]:
(12/67)

0.1791044776119403

In [147]:
#trying to select the index label for P4 and column name for admit but getting an error, likely on the first argument
h = ft.at['[1, 0, 0, 0, 2]', 'admit']

KeyError: 'admit'

#### Question 3.2. Now calculate the odds of admission if you did not attend a #1 ranked college.

In [69]:
((28+53+33)/(121+148+61))

0.34545454545454546

#### Question 3.3. Calculate the odds ratio.

In [78]:
((55*114)/(12*216))

2.4189814814814814

#### Question 3.4. Write this finding in a sentence:

#### Answer: A one unit increase in prestige decreases the probability that an applicant will be admitted to grad school at UCLA by xx%

#### Question 3.5. Print the cross tab for prestige_4.

In [72]:
pd.crosstab(index = handCalc['Prestige_4.0'], columns = handCalc['admit'])

admit           0    1
Prestige_4.0          
0.0           216  114
1.0            55   12

#### Question 3.6. Calculate the OR.

In [76]:
(216*12)/(114*55)

0.4133971291866029

#### Question 3.7. Write this finding in a sentence.

#### Answer: Increasing from Prestige 3 to Prestige 4 (increases/decreases) the probability of admission to grad school at UCLA by xx%

## Part 4. Analysis

In [83]:
# create a clean data frame for the regression
cols_to_keep = ['admit', 'gre', 'gpa']
data = df[cols_to_keep].join(prestige_dummy_df.ix[:, 'Prestige_1.0':])
print(data.head())

   admit    gre   gpa  Prestige_1.0  Prestige_2.0  Prestige_3.0  Prestige_4.0
0      0  380.0  3.61           0.0           0.0           1.0           0.0
1      1  660.0  3.67           0.0           0.0           1.0           0.0
2      1  800.0  4.00           1.0           0.0           0.0           0.0
3      1  640.0  3.19           0.0           0.0           0.0           1.0
4      0  520.0  2.93           0.0           0.0           0.0           1.0


We're going to add a constant term for our Logistic Regression.  The statsmodels function we're going to be using requires that intercepts/constants are specified explicitly.

In [ ]:
# manually add the intercept
data['intercept'] = 1.0

#### Question 4.1. Set the covariates to a variable called train_cols.

In [148]:
train_cols = data[['gre', 'gpa', 'Prestige_1.0', 'Prestige_2.0', 'Prestige_3.0', 'Prestige_4.0']]

#### Question 4.2. Fit the model.

In [149]:
logit = sm.Logit(data['admit'], train_cols)
logit_result = logit.fit()

Optimization terminated successfully.
         Current function value: 0.573854
         Iterations 6


#### Question 4.3. Print the summary results.

In [93]:
logit_result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:                  admit   No. Observations:                  397
Model:                          Logit   Df Residuals:                      391
Method:                           MLE   Df Model:                            5
Date:                Mon, 06 Jun 2016   Pseudo R-squ.:                 0.08166
Time:                        14:15:10   Log-Likelihood:                -227.82
converged:                       True   LL-Null:                       -248.08
                                        LLR p-value:                 1.176e-07
================================================================================
                   coef    std err          z      P>|z|      [95.0% Conf. Int.]
--------------------------------------------------------------------------------
gre              0.0022      0.001      2.028      0.043      7.44e-05     0.004
gpa              0.7793      0.333      2.344      0.019         0.128     1.431
Prestige_1.0    -3.8769      1.142     -3.393      0.001        -6.116    -1.638
Prestige_2.0    -4.5570      1.113     -4.093      0.000        -6.739    -2.375
Prestige_3.0    -5.2155      1.151     -4.530      0.000        -7.472    -2.959
Prestige_4.0    -5.4303      1.140     -4.764      0.000        -7.664    -3.196
================================================================================
"""

#### Question 4.4. Calculate the odds ratios of the coefficients and their 95% CI intervals

hint 1: np.exp(X)

hint 2: conf['OR'] = params

        conf.columns = ['2.5%', '97.5%', 'OR']

In [98]:
np.exp(logit_result.params)

gre             1.002221
gpa             2.180027
Prestige_1.0    0.020716
Prestige_2.0    0.010494
Prestige_3.0    0.005432
Prestige_4.0    0.004382
dtype: float64

In [156]:
conf.columns = ['2.5%', '97.5%', 'OR']

NameError: name 'conf' is not defined

In [ ]:
conf['OR'] = np.exp(logit_result.params)

#### Question 4.5. Interpret the OR of Prestige_2.

Answer:

#### Question 4.6. Interpret the OR of GPA.

Answer:

## Part 5: Predicted probablities


As a way of evaluating our classifier, we're going to recreate the dataset with every logical combination of input values.  This will allow us to see how the predicted probability of admission increases/decreases across different variables.  First we're going to generate the combinations using a helper function called cartesian (above).

We're going to use np.linspace to create a range of values for "gre" and "gpa".  This creates a range of linearly spaced values from a specified min and maximum value--in our case just the min/max observed values.

In [162]:
def cartesian(arrays, out=None):
    """
    Generate a cartesian product of input arrays.
    Parameters
    ----------
    arrays : list of array-like
        1-D arrays to form the cartesian product of.
    out : ndarray
        Array to place the cartesian product in.
    Returns
    -------
    out : ndarray
        2-D array of shape (M, len(arrays)) containing cartesian products
        formed of input arrays.
    Examples
    --------
    >>> cartesian(([1, 2, 3], [4, 5], [6, 7]))
    array([[1, 4, 6],
           [1, 4, 7],
           [1, 5, 6],
           [1, 5, 7],
           [2, 4, 6],
           [2, 4, 7],
           [2, 5, 6],
           [2, 5, 7],
           [3, 4, 6],
           [3, 4, 7],
           [3, 5, 6],
           [3, 5, 7]])
    """

    arrays = [np.asarray(x) for x in arrays]
    dtype = arrays[0].dtype

    n = np.prod([x.size for x in arrays])
    if out is None:
        out = np.zeros([n, len(arrays)], dtype=dtype)

    m = n / arrays[0].size
    out[:,0] = np.repeat(arrays[0], m)
    if arrays[1:]:
        cartesian(arrays[1:], out=out[0:m,1:])
        for j in xrange(1, arrays[0].size):
            out[j*m:(j+1)*m,1:] = out[0:m,1:]
    return out

In [161]:
# instead of generating all possible values of GRE and GPA, we're going
# to use an evenly spaced range of 10 values from the min to the max
gres = np.linspace(data['gre'].min(), data['gre'].max(), 10)

print(gres)
# array([ 220.        ,  284.44444444,  348.88888889,  413.33333333,
#         477.77777778,  542.22222222,  606.66666667,  671.11111111,
#         735.55555556,  800.        ])

gpas = np.linspace(data['gpa'].min(), data['gpa'].max(), 10)

print(gpas)
# array([ 2.26      ,  2.45333333,  2.64666667,  2.84      ,  3.03333333,
#         3.22666667,  3.42      ,  3.61333333,  3.80666667,  4.        ])

# enumerate all possibilities
combos = pd.DataFrame(cartesian([gres, gpas, [1, 2, 3, 4], [1.]]))

[ 220.          284.44444444  348.88888889  413.33333333  477.77777778
  542.22222222  606.66666667  671.11111111  735.55555556  800.        ]
[ 2.26        2.45333333  2.64666667  2.84        3.03333333  3.22666667
  3.42        3.61333333  3.80666667  4.        ]


//anaconda/lib/python3.5/site-packages/ipykernel/__main__.py:42: DeprecationWarning: using a non-integer number instead of an integer will result in an error in the future


NameError: name 'xrange' is not defined

#### Question 5.1. Recreate the dummy variables.

In [ ]:
# recreate the dummy variables

# keep only what we need for making predictions


#### Question 5.2. Make predictions on the enumerated dataset.

#### Question 5.3. Interpret findings for the last 4 observations.

Answer:

## Bonus

Plot the probability of being admitted into graduate school, stratified by GPA and GRE score.